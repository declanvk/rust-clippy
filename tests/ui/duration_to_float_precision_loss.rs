#![warn(clippy::duration_to_float_precision_loss)]

use std::time::Duration;

const DURATION: Duration = Duration::from_nanos(1_234_567_890);

fn main() {
    let float_secs = DURATION.as_secs() as f64;

    macro_rules! m {
        ($d:expr) => {{
            let float_secs = $d.as_secs() as f64;
        }};
    }

    m!(DURATION);

    let float_secs = DURATION.as_secs() as f32;

    let float_secs = DURATION.as_millis() as f64 / 1_000.0;

    let float_secs = DURATION.as_micros() as f64 / 1_000_000.0;

    let float_secs = DURATION.as_micros() as f64 * 1e-6;

    let float_secs = DURATION.as_nanos() as f64 / 1_000_000_000.0;

    let float_millis = DURATION.as_millis() as f64;

    let float_micros = DURATION.as_micros() as f64;

    // should not trigger on these conversion, since they are already the max precision
    let float_nanos = DURATION.as_nanos() as f64;

    let float_micros = DURATION.as_nanos() as f64 / 1_000.0;

    let float_millis = DURATION.as_nanos() as f64 / 1_000_000.0;

    {
        let dur = &&DURATION;
        let float_millis = dur.as_millis() as f64 / 1_000.0;
    }
}

#[clippy::msrv = "1.37"]
fn msrv_1_37() {
    let float_secs = DURATION.as_secs() as f64; // should not trigger lint because MSRV is below
}

#[clippy::msrv = "1.38"]
fn msrv_1_38() {
    let float_secs = DURATION.as_secs() as f64; // should trigger lint because MSRV is satisfied
}
