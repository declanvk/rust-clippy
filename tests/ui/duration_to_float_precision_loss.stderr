error: calling `as_secs_f64()` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:8:22
   |
LL |     let float_secs = DURATION.as_secs() as f64;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f64()`
   |
   = note: `-D clippy::duration-to-float-precision-loss` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::duration_to_float_precision_loss)]`

error: calling `as_secs_f32()` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:18:22
   |
LL |     let float_secs = DURATION.as_secs() as f32;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f32()`

error: calling `as_secs_f64()` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:20:22
   |
LL |     let float_secs = DURATION.as_millis() as f64 / 1_000.0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f64()`

error: calling `as_secs_f64()` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:22:22
   |
LL |     let float_secs = DURATION.as_micros() as f64 / 1_000_000.0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f64()`

error: calling `as_secs_f64()` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:24:22
   |
LL |     let float_secs = DURATION.as_micros() as f64 * 1e-6;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f64()`

error: calling `as_secs_f64()` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:26:22
   |
LL |     let float_secs = DURATION.as_nanos() as f64 / 1_000_000_000.0;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f64()`

error: calling `as_secs_f64() * 1e3` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:28:24
   |
LL |     let float_millis = DURATION.as_millis() as f64;
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f64() * 1e3`

error: calling `as_secs_f64() * 1e6` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:30:24
   |
LL |     let float_micros = DURATION.as_micros() as f64;
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f64() * 1e6`

error: calling `as_secs_f64()` is more precise than this calculation
  --> tests/ui/duration_to_float_precision_loss.rs:47:22
   |
LL |     let float_secs = DURATION.as_secs() as f64; // should trigger lint because MSRV is satisfied
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `DURATION.as_secs_f64()`

error: aborting due to 9 previous errors

